# <img style="float: left; padding-right: 20px; width: 200px" src="https://raw.githubusercontent.com/raxlab/imt2200-data/main/media/logo.jpg">  IMT 2200 - Introducción a Ciencia de Datos
**Pontificia Universidad Católica de Chile**<br>
**Instituto de Ingeniería Matemática y Computacional**<br>
**Semestre 2025-S2**<br>
**Profesor:** Rodrigo A. Carrasco <br>

# <h1><center>Actividad 7 - Actividad de Integración</center></h1>

Esta actividad busca que los estudiantes conecten el trabajo de agrupación y de resumen, junto con calcular estadísticas de datos así como también realizar visualización básica.

## Instrucciones

Este Notebook contiene las instrucciones a realizar para la actividad. 

<b>Al finalizarla, deben subir el Notebook y los archivos generados en un único archivo .zip, al módulo de la Actividad 07 en Canvas.</b>

Para esta actividad trabajarán con el dataset [palmerpenguins](https://github.com/allisonhorst/palmerpenguins/blob/main/README.md), un set de datos con mediciones de 3 especies de pingüinos que habitan en la Antártica y sus islas. Estos datos incluyen peso, tamaño de las aletas y mediciones del "culmen" para hembras y machos de las 3 especies. (Abajo pueden ver una imagen explicativa sobre estas mediciones.)

<img src="culmen_depth.png" alt="bill_size" width="700"/>

## Rúbrica

- Si han logrado todo: 7.0
- Si han logrado sólo los puntos 1 y 2: 5.0
- Menos que eso: 1.0

## 1. Base de datos unificada

### 1.1 Lectura de datos

En la carpeta `penguins/` podrá encontrar tres archivos `.csv`, cada uno con registros de las mediciones de una especie de pingüino antártico.
Primero, lea cada uno de estos archivos como un DataFrame separado.

### 1.2  Base de datos unificada

Ahora junte los tres DataFrames en un nuevo DataFrame llamado `penguins`, agregando una nueva columna `species` que identifique el nombre de la especie de cada pingüino (Adelie, Chinstrap o Gentoo).

### 1.3 Inspección y limpieza

Inspeccione el DataFrame `penguins` y sus columnas. ¿Están correctamente asignados los tipos de cada columna? ¿Alguna podría ser categórica? ¿Existen valores nulos o inválidos? Aplique las transformaciones y limpiezas que considere necesarisas al DataFrame, explicando en los comentarios (o nuevas celdas) por qué hizo dichas transformaciones.

In [2]:
import pandas as pd


In [3]:
df1 = pd.read_csv('penguins/adelie.csv')
df2 = pd.read_csv('penguins/chinstrap.csv')
df3 = pd.read_csv('penguins/gentoo.csv')

In [10]:
display(df1.head(), df2.head(), df3.head())

,island,culmen_length_mm,culmen_depth_mm,flipper_length_mm,body_mass_g,sex
0,Torgersen,39.1,18.7,181.0,3750.0,MALE
1,Torgersen,39.5,17.4,186.0,3800.0,FEMALE
2,Torgersen,40.3,18.0,195.0,3250.0,FEMALE
3,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Torgersen,36.7,19.3,193.0,3450.0,FEMALE


,island,culmen_length_mm,culmen_depth_mm,flipper_length_mm,body_mass_g,sex
0,Dream,46.5,17.9,192.0,3500.0,FEMALE
1,Dream,50.0,19.5,196.0,3900.0,MALE
2,Dream,51.3,19.2,193.0,3650.0,MALE
3,Dream,45.4,18.7,188.0,3525.0,FEMALE
4,Dream,52.7,19.8,197.0,3725.0,MALE


,island,culmen_length_mm,culmen_depth_mm,flipper_length_mm,body_mass_g,sex
0,Biscoe,46.1,13.2,211.0,4500.0,FEMALE
1,Biscoe,50.0,16.3,230.0,5700.0,MALE
2,Biscoe,48.7,14.1,210.0,4450.0,FEMALE
3,Biscoe,50.0,15.2,218.0,5700.0,MALE
4,Biscoe,47.6,14.5,215.0,5400.0,MALE


In [ ]:
#Agrego las especies en cada df por facilidad
df1['species'] = 'Adelie'
df2['species'] = 'Chinstrap'
df3['species'] = 'Gentoo'

In [13]:
penguins = pd.concat([df1, df2, df3])

In [15]:
display(penguins.head(), penguins.tail())

,island,culmen_length_mm,culmen_depth_mm,flipper_length_mm,body_mass_g,sex,species
0,Torgersen,39.1,18.7,181.0,3750.0,MALE,Adelie
1,Torgersen,39.5,17.4,186.0,3800.0,FEMALE,Adelie
2,Torgersen,40.3,18.0,195.0,3250.0,FEMALE,Adelie
3,Torgersen,NaN,NaN,NaN,NaN,NaN,Adelie
4,Torgersen,36.7,19.3,193.0,3450.0,FEMALE,Adelie


,island,culmen_length_mm,culmen_depth_mm,flipper_length_mm,body_mass_g,sex,species
119,Biscoe,NaN,NaN,NaN,NaN,NaN,Gentoo
120,Biscoe,46.8,14.3,215.0,4850.0,FEMALE,Gentoo
121,Biscoe,50.4,15.7,222.0,5750.0,MALE,Gentoo
122,Biscoe,45.2,14.8,212.0,5200.0,FEMALE,Gentoo
123,Biscoe,49.9,16.1,213.0,5400.0,MALE,Gentoo


In [20]:
#1.3
print("\nValores nulos por columna:")
penguins.isna().sum()


Valores nulos por columna:


island                0
culmen_length_mm      2
culmen_depth_mm       2
flipper_length_mm     2
body_mass_g           2
sex                  10
species               0
dtype: int64

In [23]:
print("Tipos de datos antes de la limpieza:")
print(penguins.dtypes)

Tipos de datos antes de la limpieza:
island               category
culmen_length_mm      float64
culmen_depth_mm       float64
flipper_length_mm     float64
body_mass_g           float64
sex                  category
species              category
dtype: object


In [ ]:
penguins['species'] = penguins['species'].astype('category')
penguins['sex'] = penguins['sex'].astype('category')
penguins['island'] = penguins['island'].astype('category')

#Elimino filas nulas
penguins_cleaned = penguins.dropna()

# Verificar los tipos de datos después de la limpieza
print("\nTipos de datos después de la limpieza:")
print(penguins_cleaned.dtypes)


Tipos de datos antes de la limpieza:
island               category
culmen_length_mm      float64
culmen_depth_mm       float64
flipper_length_mm     float64
body_mass_g           float64
sex                  category
species              category
dtype: object

Tipos de datos después de la limpieza:
island               category
culmen_length_mm      float64
culmen_depth_mm       float64
flipper_length_mm     float64
body_mass_g           float64
sex                  category
species              category
dtype: object


In [ ]:
print("\nPrimeras filas del DataFrame limpio:")
print(penguins_cleaned.head())

## 2. Revisión de base de datos

### 2.1 Análisis de base completa

Responda las siguientes preguntas a partir del DataFrame `penguins`:
* a. ¿Cuántas filas tiene la base de datos conjunta?
* b. ¿Cuáles son los valores mínimo y máximo de todas las columnas con datos numéricos?
* c. ¿Cuántos machos y cuántas hembras hay en el set de datos, respectivamente?

### 2.2 Análisis agrupados

Responda las siguientes preguntas:

* a. En promedio, ¿cuál es la especie más pesada de pingüino?
* b. Para cada especie, ¿hay alguna diferencia significativa entre hembras y machos en relación al peso promedio?
* c. ¿En qué especie varía menos el tamaño de aleta?

In [40]:
print('2.1A')
print(f'Número de entradas del dataframe limpio: {len(penguins_cleaned)}')
print('\n2.1B')
display(penguins_cleaned.describe().round(2))

print('\n2.1C')
nro_machos = len(penguins_cleaned[penguins_cleaned['sex'] == 'MALE'])
nro_hembras = len(penguins_cleaned[penguins_cleaned['sex'] == 'FEMALE'])
print(f"Machos: {nro_machos}")
print(f"Hembras: {nro_hembras}")

2.1A
Número de entradas del dataframe limpio: 334

2.1B


,culmen_length_mm,culmen_depth_mm,flipper_length_mm,body_mass_g
count,334.00,334.00,334.00,334.00
mean,43.99,17.16,201.01,4209.06
std,5.46,1.97,14.02,804.84
min,32.10,13.10,172.00,2700.00
25%,39.50,15.60,190.00,3550.00
50%,44.50,17.30,197.00,4050.00
75%,48.58,18.70,213.00,4793.75
max,59.60,21.50,231.00,6300.00



2.1C
Machos: 168
Hembras: 165


In [45]:
print('2.2A')
promedios_peso = penguins_cleaned.groupby('species')['body_mass_g'].mean()


especie_mas_pesada = promedios_peso.idxmax()
mayor_peso = promedios_peso.max()

print("Peso promedio por especie:")
print(promedios_peso)
print(f"\nLa especie más pesada en promedio es {especie_mas_pesada} con un peso promedio de {mayor_peso} g.")

2.2A
Peso promedio por especie:
species
Adelie       3706.164384
Chinstrap    3733.088235
Gentoo       5090.625000
Name: body_mass_g, dtype: float64

La especie más pesada en promedio es Gentoo con un peso promedio de 5090.625 g.


/var/folders/hf/z4nqpl692qq20c0yrl9q93r40000gn/T/ipykernel_3458/3394160408.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  promedios_peso = penguins_cleaned.groupby('species')['body_mass_g'].mean()


In [ ]:
prom_especie_sexo = penguins_cleaned.groupby(['species', 'sex'])['body_mass_g'].mean()

print("Peso promedio por especie y sexo:")
print(prom_especie_sexo)

print("\n"+"="*50)
#Comparo diferencias entre especies y sexo
for species in penguins_cleaned['species'].unique():
    peso_macho = prom_especie_sexo[species]['MALE']
    peso_hembra = prom_especie_sexo[species]['FEMALE']
    diff = abs(peso_macho - peso_hembra)
    print(f"\nEn la especie {species}:")
    print(f"- Machos: {peso_macho:.2f} g")
    print(f"- Hembras: {peso_hembra} g")
    print(f"- Diferencia: {diff.round(2)} g")

Peso promedio por especie y sexo:
species    sex   
Adelie     .                 NaN
           FEMALE    3368.835616
           MALE      4043.493151
Chinstrap  .                 NaN
           FEMALE    3527.205882
           MALE      3938.970588
Gentoo     .         4875.000000
           FEMALE    4679.741379
           MALE      5484.836066
Name: body_mass_g, dtype: float64


En la especie Adelie:
- Machos: 4043.49 g
- Hembras: 3368.8356164383563 g
- Diferencia: 674.66 g

En la especie Chinstrap:
- Machos: 3938.97 g
- Hembras: 3527.205882352941 g
- Diferencia: 411.76 g

En la especie Gentoo:
- Machos: 5484.84 g
- Hembras: 4679.741379310345 g
- Diferencia: 805.09 g


/var/folders/hf/z4nqpl692qq20c0yrl9q93r40000gn/T/ipykernel_3458/1264887957.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  prom_especie_sexo = penguins_cleaned.groupby(['species', 'sex'])['body_mass_g'].mean()


In [58]:
print(f"Diferencia promedio: {round((674.66 + 411.76 + 805.09)/3, 2)}")

Diferencia promedio: 630.5


Si existe una diferencia significativa, los machos tienden a tener un peso mayor a las hembras con una diferencia promedio de 630.5 gramos 

In [64]:
#Para la variacion, uso std (desviacion estandard)
std_ala_especie = penguins_cleaned.groupby('species')['flipper_length_mm'].std()

# Encontrar la especie con menor variación
especie_menor_variacion = std_ala_especie.idxmin()
least_variation = std_ala_especie.min()

print("\nDesviación estándar del tamaño de aleta por especie:")
print(std_ala_especie)
print(f"\nLa especie con menor variación en el tamaño de aleta es {especie_menor_variacion} con una desviación estándar de {round(least_variation, 2)} mm.")


Desviación estándar del tamaño de aleta por especie:
species
Adelie       6.521825
Chinstrap    7.131894
Gentoo       6.557738
Name: flipper_length_mm, dtype: float64

La especie con menor variación en el tamaño de aleta es Adelie con una desviación estándar de 6.52 mm.


/var/folders/hf/z4nqpl692qq20c0yrl9q93r40000gn/T/ipykernel_3458/4037956139.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  std_ala_especie = penguins_cleaned.groupby('species')['flipper_length_mm'].std()


## 3. Análisis exploratorio

### 3.1 Estadísticas de resumen

Elija una de las tres especies y responda las siguientes preguntas solamente para la especie elegida:

* a. ¿Cuál es el coeficiente de correlación entre `culmen_length` y `flipper_length`? ¿Qué nos indica este valor? Comente.
* b. ¿Cual es el peso máximo del percentil 10 (el decil más bajo) para los machos?
* c. ¿Cuál es la varianza de `culmen_depth` para las hembras?

### 3.2 Visualización

Utilizando la librería de visualización de su preferencia (matplotlib, seaborn, plotly), genere los siguientes gráficos:

* a. Un scatterplot que relacione las variables `culmen_length` y `flipper_length` para las tres especies. Utilice colores distintos para diferenciar las especies entre sí.
* b. Un histograma que muestre la distribución de peso para pingüinos macho y hembra. Utilice colores distintos para diferenciar los dos grupos.

Analice y comente los resultados que obtuvo. No olvide colocar título y nombres claros a los ejes de sus gráficos.